File: app/core/abstract_script.py
# -*- coding: utf-8 -*-
"""Абстрактный базовый класс для всех скриптов обработки."""

import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
import threading
from typing import Any, Callable


logger = logging.getLogger(__name__)


class SettingType(Enum):
    """Типы полей настроек скрипта."""

    TEXT = "text"
    COMBO = "combo"
    CHECKBOX = "checkbox"


@dataclass(frozen=True)
class SettingField:
    """Описание одного поля настроек скрипта.

    Attributes:
        key: Уникальный ключ настройки.
        label: Отображаемое название настройки.
        setting_type: Тип виджета настройки.
        default: Значение по умолчанию.
        options: Варианты выбора для COMBO.
    """

    key: str
    label: str
    setting_type: SettingType
    default: Any = ""
    options: list[str] = field(default_factory=list)
    visible_if: dict[str, list[Any]] = field(default_factory=dict)


# Тип callback-функции для отчёта о прогрессе.
ProgressCallback = Callable[[int, int, str], None]


class AbstractScript(ABC):
    """Абстрактный базовый класс скрипта обработки файлов.

    Определяет контракт, которому должны следовать
    все конкретные реализации скриптов. Каждый скрипт
    предоставляет метаданные (имя, описание, расширения,
    схему настроек) и метод выполнения.
    """

    @property
    @abstractmethod
    def name(self) -> str:
        """Отображаемое имя скрипта."""

    @property
    @abstractmethod
    def category(self) -> str:
        """Категория скрипта (например, 'Аудио', 'Видео')."""

    @property
    @abstractmethod
    def description(self) -> str:
        """Краткое описание скрипта для UI."""

    @property
    @abstractmethod
    def icon_name(self) -> str:
        """Имя иконки из FluentIcon (например, 'VIDEO')."""

    @property
    @abstractmethod
    def file_extensions(self) -> list[str]:
        """Допустимые расширения входных файлов.

        Примеры: ['.mp4', '.mkv', '.avi'].
        """

    @property
    def settings_schema(self) -> list[SettingField]:
        """Схема настроек скрипта.

        По умолчанию возвращает пустой список.
        Переопредели в подклассе для добавления настроек.
        """
        return []

    @property
    def use_custom_widget(self) -> bool:
        """Использовать ли кастомный виджет файлов.

        Если True, WorkPanel должна использовать специфичный
        виджет (например, таблицу муксинга) вместо стандартного
        списка файлов.
        """
        return False

    @property
    def supports_parallel(self) -> bool:
        """Поддерживает ли скрипт параллельную обработку файлов.

        Если True, ScriptWorker может запускать обработку
        нескольких файлов одновременно в разных потоках.
        """
        return False

    @property
    def is_cancelled(self) -> bool:
        """Проверить, был ли скрипт отменен."""
        return getattr(self, "_is_cancelled", False)

    def cancel(self) -> None:
        """Отменить выполнение скрипта."""
        self._is_cancelled = True
        logger.info("Скрипт '%s' получил команду на отмену.", self.name)

    def _delete_source(
        self,
        file_path: Path,
        results: list[str],
    ) -> None:
        """Удалить исходный файл после обработки.

        Args:
            file_path: Путь к удаляемому файлу.
            results: Список результатов для лога.
        """
        try:
            file_path.unlink()
            msg = f"🗑 Удалён исходник: {file_path.name}"
            results.append(msg)
            logger.info(msg)
        except OSError:
            logger.exception(
                "Не удалось удалить файл '%s'",
                file_path.name,
            )
            results.append(f"⚠ Не удалось удалить: " f"{file_path.name}")

    def _replace_source_with_result(
        self,
        source_path: Path,
        result_path: Path,
        results: list[str],
    ) -> bool:
        """Заменить исходный файл результатом (физическая подмена).

        Этот метод удаляет оригинал и переименовывает результат работы скрипта
        так, чтобы он занял место оригинала с сохранением исходного имени.

        Returns:
            True, если замена прошла успешно.
        """
        try:
            # 1. Удаляем оригинал
            source_path.unlink()
            # 2. Переименовываем результат в имя оригинала
            result_path.rename(source_path)

            msg = f"🔄 Подменен оригинал: {source_path.name}"
            results.append(msg)
            logger.info(msg)
            return True
        except OSError as e:
            logger.exception(
                "Ошибка при подмене файла '%s': %s",
                source_path.name,
                e,
            )
            results.append(f"❌ Ошибка подмены: {source_path.name}")
            return False

    def _cleanup_if_cancelled(self, *paths: Path) -> None:
        """Удалить неполные выходные файлы при отмене работы скрипта."""
        if not self.is_cancelled:
            return

        for path in paths:
            if path and path.exists():
                try:
                    path.unlink(missing_ok=True)
                    logger.debug(
                        "🗑 Удален неполный выходной файл: %s", path.name
                    )
                except Exception as e:
                    logger.warning("⚠ Не удалось удалить %s: %s", path.name, e)

    def _get_safe_output_path(
        self,
        input_path: Path,
        output_path: Path,
    ) -> Path:
        """Получить безопасный путь для выходного файла.

        Защищает исходный файл от перезаписи (добавляет '_processed').
        """
        try:
            # Используем resolve() и сравнение строк в нижнем регистре для Windows  # noqa: E501
            in_resolved = str(input_path.resolve()).lower()
            out_resolved = str(output_path.resolve()).lower()

            logger.debug(
                "Сравнение путей: IN=%s, OUT=%s", in_resolved, out_resolved
            )

            # 1. Если выход совпадает с исходником — ОБЯЗАТЕЛЬНО меняем имя
            if in_resolved == out_resolved:
                output_path = (
                    output_path.parent
                    / f"{output_path.stem}_processed{output_path.suffix}"
                )
                logger.info(
                    "Защита исходника: добавлено '_processed' к имени (%s)",
                    output_path.name,
                )

            # 2. Защита от коллизий имен файлов при пакетной обработке
            if not hasattr(self, "_batch_reserved_paths"):
                self.prepare_batch()

            with self._batch_lock:
                original_stem = output_path.stem
                counter = 1
                while (
                    str(output_path.resolve()).lower()
                    in self._batch_reserved_paths
                ):
                    output_path = (
                        output_path.parent
                        / f"{original_stem}_{counter}{output_path.suffix}"
                    )
                    counter += 1

                self._batch_reserved_paths.add(
                    str(output_path.resolve()).lower()
                )

        except Exception as exc:
            logger.exception("Ошибка в _get_safe_output_path: %s", exc)

        return output_path

    def prepare_batch(self, input_files: list[Path] | None = None) -> None:
        """Подготовка к пакетной обработке.

        Очищает список зарезервированных путей для выходных файлов
        и добавляет входные пути пакета для предотвращения их перезаписи.
        Должен быть вызван перед началом обработки новой партии файлов.
        """
        if not hasattr(self, "_batch_reserved_paths"):
            self._batch_reserved_paths: set[str] = set()
            self._batch_lock = threading.Lock()
        with self._batch_lock:
            self._batch_reserved_paths.clear()
            if input_files:
                for f in input_files:
                    self._batch_reserved_paths.add(str(f.resolve()).lower())
        logger.debug(
            "Инициализирован список зарезервированных путей "
            "для пакетной обработки."
        )

    def execute(
        self,
        files: list[Path],
        settings: dict[str, Any],
        output_path: str | None = None,
        progress_callback: ProgressCallback | None = None,
    ) -> list[str]:
        """Выполнить обработку списка файлов (Template Method).

        Обеспечивает базовую логику цикла обработки. Скрипты, требующие
        группировки файлов (например, Муксер), переопределяют этот метод.
        Все остальные реализуют execute_single.

        Args:
            files: Список путей к входным файлам.
            settings: Словарь текущих настроек скрипта.
            output_path: Опциональный путь сохранения (ручной выбор).
            progress_callback: Callback (текущий, всего, сообщение).

        Returns:
            Список строк-результатов выполнения для лога.
        """
        results: list[str] = []
        total = len(files)

        for i, file_path in enumerate(files):
            if self.is_cancelled:
                msg = "⚠ Обработка прервана пользователем."
                logger.info(msg)
                results.append(msg)
                break

            if progress_callback:
                progress_callback(i, total, f"Обработка: {file_path.name}")

            # Обработка одного файла
            try:
                res = self.execute_single(file_path, settings, output_path)
                results.extend(res)
            except Exception as e:
                msg = f"❌ Критическая ошибка при обработке {file_path.name}: {e}"  # noqa: E501
                logger.exception(msg)
                results.append(msg)

            if progress_callback:
                # Показываем результат последнего обработанного файла в статусе
                status_msg = results[-1] if results else ""
                progress_callback(i + 1, total, status_msg)

        return results

    @abstractmethod
    def execute_single(
        self,
        file: Path,
        settings: dict[str, Any],
        output_path: str | None = None,
    ) -> list[str]:
        """Обработать один файл.

        Должен быть реализован во всех скриптах, поддерживающих
        пофайловую или параллельную обработку.

        Args:
            file: Путь к входному файлу.
            settings: Словарь настроек.
            output_path: Опциональный путь сохранения.

        Returns:
            Список строк-результатов.
        """

File: src-csharp/KTools.App/Core/AbstractScript.cs
// -*- coding: utf-8 -*-
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text.Json;
using System.Threading.Tasks;

namespace KTools_App.Core;

/// <summary>
/// Абстрактный базовый класс для всех скриптов обработки файлов в K-Tools.
/// Определяет контракт метаданных, зависимостей и асинхронного выполнения.
/// </summary>
public abstract class AbstractScript
{
    private volatile bool _isCancelled;

    /// <summary>
    /// Отображаемое на русском языке имя скрипта.
    /// </summary>
    public abstract string Name { get; }

    /// <summary>
    /// Описание назначения скрипта и его ключевых особенностей.
    /// </summary>
    public abstract string Description { get; }

    /// <summary>
    /// Категория скрипта (например, "Видео", "Аудио", "Контейнеры", "Субтитры").
    /// </summary>
    public abstract string Category { get; }

    /// <summary>
    /// Имя иконки для Fluent-отображения.
    /// </summary>
    public abstract string IconName { get; }

    /// <summary>
    /// Список поддерживаемых расширений файлов в нижнем регистре с точкой (например, .mkv, .mp4).
    /// </summary>
    public abstract string[] FileExtensions { get; }

    /// <summary>
    /// Декларативная схема параметров настроек скрипта.
    /// По умолчанию возвращает пустой список.
    /// </summary>
    public virtual List<SettingField> SettingsSchema => new();

    /// <summary>
    /// Указывает, поддерживает ли скрипт одновременную параллельную обработку файлов.
    /// </summary>
    public virtual bool SupportsParallel => false;

    /// <summary>
    /// Указывает, использует ли скрипт кастомный виджет выбора дорожек (например, TreeView).
    /// </summary>
    public virtual bool UseCustomWidget => false;

    /// <summary>
    /// Список строковых ключей внешних зависимостей, необходимых для работы скрипта.
    /// </summary>
    public virtual string[] RequiredDependencies => Array.Empty<string>();

    /// <summary>
    /// Проверить, была ли отправлена команда отмены выполнения скрипта.
    /// </summary>
    public bool IsCancelled => _isCancelled;

    /// <summary>
    /// Инициировать отмену выполнения текущего скрипта.
    /// </summary>
    public virtual void Cancel()
    {
        _isCancelled = true;
    }

    /// <summary>
    /// Сбросить состояние отмены перед новым запуском пакета файлов.
    /// </summary>
    public virtual void ResetCancellation()
    {
        _isCancelled = false;
    }

    /// <summary>
    /// Инициализирует новый экземпляр класса <see cref="AbstractScript"/>
    /// и настраивает автоматическое удаление сохраненного выбора при удалении файлов из очереди.
    /// </summary>
    protected AbstractScript()
    {
        FilesQueue.CollectionChanged += (sender, e) =>
        {
            if (e.Action == System.Collections.Specialized.NotifyCollectionChangedAction.Remove && e.OldItems != null)
            {
                foreach (FileQueueItem item in e.OldItems)
                {
                    SelectedTrackIds.Remove(item.FilePath);
                    SelectedAttachmentIds.Remove(item.FilePath);
                    LogService.Instance.DebugLog(
                        $"Очищен сохраненный выбор дорожек для удаленного файла: '{item.FileName}'", 
                        "AbstractScript");
                }
            }
            else if (e.Action == System.Collections.Specialized.NotifyCollectionChangedAction.Reset)
            {
                SelectedTrackIds.Clear();
                SelectedAttachmentIds.Clear();
                LogService.Instance.DebugLog(
                    "Очищен весь сохраненный выбор дорожек в связи со сбросом очереди файлов", 
                    "AbstractScript");
            }
        };
    }

    /// <summary>
    /// Словарь выбранных дорожек для файлов (путь -> список ID дорожек).
    /// </summary>
    public Dictionary<string, List<int>> SelectedTrackIds { get; } = new();

    /// <summary>
    /// Словарь выбранных вложений для файлов (путь -> список ID вложений).
    /// </summary>
    public Dictionary<string, List<int>> SelectedAttachmentIds { get; } = new();

    /// <summary>
    /// Очередь файлов скрипта, сохраняющаяся между переходами.
    /// </summary>
    public ObservableCollection<FileQueueItem> FilesQueue { get; } = new();

    /// <summary>
    /// Сохраненный текст журнала выполнения скрипта.
    /// </summary>
    public string SavedLogText { get; set; } = string.Empty;

    /// <summary>
    /// Сохраненный глобальный текстовый статус медиаобработки.
    /// </summary>
    public string SavedStatusText { get; set; } = "Ожидание запуска...";

    /// <summary>
    /// Сохраненное значение интегрального прогресс-бара.
    /// </summary>
    public double SavedGlobalProgress { get; set; }

    /// <summary>
    /// Указывает, выполняется ли скрипт в данный момент.
    /// </summary>
    public bool IsProcessing { get; set; }

    /// <summary>
    /// Событие изменения состояния выполнения скрипта.
    /// </summary>
    public event EventHandler? StateChanged;

    /// <summary>
    /// Вызывает событие изменения состояния для подписчиков.
    /// </summary>
    public void RaiseStateChanged()
    {
        StateChanged?.Invoke(this, EventArgs.Empty);
    }

    private readonly object _batchLock = new();
    private readonly HashSet<string> _batchReservedPaths = new(StringComparer.OrdinalIgnoreCase);

    /// <summary>
    /// Очищает список зарезервированных путей перед началом новой пакетной обработки.
    /// </summary>
    public virtual void PrepareBatch(IEnumerable<string>? inputFiles = null)
    {
        lock (_batchLock)
        {
            _batchReservedPaths.Clear();
            if (inputFiles != null)
            {
                foreach (var file in inputFiles)
                {
                    _batchReservedPaths.Add(Path.GetFullPath(file));
                }
            }
        }
    }

    /// <summary>
    /// Возвращает безопасный путь для сохранения результата, предотвращая перезапись исходника
    /// и коллизии имен при пакетном переименовании.
    /// </summary>
    protected string GetSafeOutputPath(string inputPath, string outputPath)
    {
        try
        {
            string inResolved = Path.GetFullPath(inputPath);
            string outResolved = Path.GetFullPath(outputPath);

            // 1. Защита исходного файла от перезаписи
            if (inResolved.Equals(outResolved, StringComparison.OrdinalIgnoreCase))
            {
                string dir = Path.GetDirectoryName(outResolved) ?? "";
                string stem = Path.GetFileNameWithoutExtension(outResolved);
                string ext = Path.GetExtension(outResolved);
                outResolved = Path.Combine(dir, $"{stem}_processed{ext}");
                LogService.Instance.Info($"Защита исходника: добавлено '_processed' к имени файла: '{outResolved}'", "AbstractScript");
            }

            // 2. Защита от коллизий имен при пакетной обработке
            lock (_batchLock)
            {
                string originalDir = Path.GetDirectoryName(outResolved) ?? "";
                string originalStem = Path.GetFileNameWithoutExtension(outResolved);
                string ext = Path.GetExtension(outResolved);
                int counter = 1;

                while (_batchReservedPaths.Contains(outResolved))
                {
                    outResolved = Path.Combine(originalDir, $"{originalStem}_{counter}{ext}");
                    counter++;
                }

                _batchReservedPaths.Add(outResolved);
            }
            
            return outResolved;
        }
        catch (Exception ex)
        {
            LogService.Instance.Exception(ex, $"Ошибка при получении безопасного выходного пути для '{inputPath}': {ex.Message}", "AbstractScript");
            return outputPath;
        }
    }

    /// <summary>
    /// Физически удаляет исходный файл с диска и заносит лог в результаты.
    /// </summary>
    protected void DeleteSource(string filePath, List<string> results)
    {
        try
        {
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
                string msg = $"🗑 Удален исходник: {Path.GetFileName(filePath)}";
                results.Add(msg);
                LogService.Instance.Info(msg, "AbstractScript");
            }
        }
        catch (Exception ex)
        {
            LogService.Instance.Exception(ex, $"Не удалось удалить исходный файл '{filePath}': {ex.Message}", "AbstractScript");
            results.Add($"⚠ Не удалось удалить: {Path.GetFileName(filePath)}");
        }
    }

    /// <summary>
    /// Физически заменяет исходный файл полученным результатом с сохранением имени оригинала.
    /// </summary>
    protected bool ReplaceSourceWithResult(string sourcePath, string resultPath, List<string> results)
    {
        try
        {
            if (File.Exists(sourcePath))
            {
                File.Delete(sourcePath);
            }
            File.Move(resultPath, sourcePath);

            string msg = $"🔄 Подменен оригинал: {Path.GetFileName(sourcePath)}";
            results.Add(msg);
            LogService.Instance.Info(msg, "AbstractScript");
            return true;
        }
        catch (Exception ex)
        {
            LogService.Instance.Exception(ex, $"Ошибка при подмене оригинального файла '{sourcePath}' результатом '{resultPath}': {ex.Message}", "AbstractScript");
            results.Add($"❌ Ошибка подмены: {Path.GetFileName(sourcePath)}");
            return false;
        }
    }

    /// <summary>
    /// Удаляет незавершенные выходные файлы при прерывании процесса.
    /// </summary>
    protected void CleanupIfCancelled(string filePath)
    {
        if (!IsCancelled) return;

        try
        {
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
                LogService.Instance.DebugLog($"Удален неполный выходной файл: '{Path.GetFileName(filePath)}'", "AbstractScript");
            }
        }
        catch (Exception ex)
        {
            LogService.Instance.Warn($"Не удалось удалить временный файл '{filePath}' при отмене: {ex.Message}", "AbstractScript");
        }
    }

    /// <summary>
    /// Асинхронный запуск обработки одного файла.
    /// </summary>
    /// <param name="filePath">Абсолютный путь к файлу.</param>
    /// <param name="settings">Словарь текущих настроек пользователя.</param>
    /// <param name="outputPath">Директория сохранения результата.</param>
    /// <param name="progressCallback">Делегат для отправки прогресса (индекс, всего, сообщение, процент).</param>
    /// <param name="fileIndex">Порядковый индекс обрабатываемого файла в очереди.</param>
    /// <param name="totalCount">Общее число файлов в очереди.</param>
    /// <returns>Список сообщений о результатах выполнения (ошибки, успехи, пути подмены).</returns>
    public abstract Task<List<string>> ExecuteSingleAsync(
        string filePath,
        Dictionary<string, object> settings,
        string? outputPath,
        Action<int, int, string, double?> progressCallback,
        int fileIndex,
        int totalCount);

    /// <summary>
    /// Безопасно извлекает значение параметра из словаря настроек.
    /// Поддерживает автоматическую конвертацию JsonElement.
    /// </summary>
    protected T GetSettingValue<T>(
        Dictionary<string, object> settings,
        string key,
        T defaultValue)
    {
        if (settings == null) return defaultValue;

        if (settings.TryGetValue(key, out var val))
        {
            try
            {
                if (val is System.Text.Json.JsonElement jsonElem)
                {
                    if (typeof(T) == typeof(bool))
                    {
                        return (T)(object)(jsonElem.ValueKind == 
                            System.Text.Json.JsonValueKind.True);
                    }
                    if (typeof(T) == typeof(int))
                    {
                        return (T)(object)jsonElem.GetInt32();
                    }
                    if (typeof(T) == typeof(string))
                    {
                        return (T)(object)jsonElem.GetString()!;
                    }
                    
                    var deserialized = jsonElem.Deserialize<T>();
                    if (deserialized != null)
                    {
                        return deserialized;
                    }
                }
                else if (val is T typedVal)
                {
                    return typedVal;
                }
                else
                {
                    return (T)Convert.ChangeType(val, typeof(T));
                }
            }
            catch (Exception)
            {
                // Игнорируем ошибки приведения
            }
        }
        return defaultValue;
    }
}
